### DISCUSSION TITLE
Clarification on frequency-averaging logic in PUSCHLSChannelEstimator

### DISCUSSION THREAD

**Discussion by LiPtP0000** [NVlabs/sionna#976] [Q&A] [answered]
Hi all,
I was going through the following code in `/phy/nr/pusch_channel_estimation.py` and had a question regarding the frequency-averaging step: 
```py
# Frequency-averaging between adjacent channel estimates
# Compute number of elements across which frequency averaging should
# be done. This includes the zeroed elements.
n = 2*self._num_cdm_groups_without_data
k = int(h_hat.shape[-1]/n) # Second dimension

# Reshape last dimension to [k, n]
h_hat = split_dim(h_hat, [k, n], 5)
cond = tf.abs(h_hat)>0 # Mask for irrelevant channel estimates
h_hat = tf.reduce_sum(h_hat, axis=-1, keepdims=True) \
        / tf.cast(2,h_hat.dtype)
h_hat = tf.repeat(h_hat, n, axis=-1)
h_hat = tf.where(cond, h_hat, 0) # Mask irrelevant channel estimates
h_hat = tf.reshape(h_hat, h_ls_shape)

# The error variance gets reduced by a factor of two
err_var /= tf.cast(2, err_var.dtype)
return h_hat, err_var
```
From what I understand, `n` represents `2 * num_cdm_groups_without_data`, so for example if `num_cdm_groups_without_data = 2`, then `n = 4`. This means the `h_hat` tensor is reshaped into `k × n` groups (e.g., [v0, v1, v2, v3]), and then a reduction is done across `n` (i.e., all 4 elements in this case). However, the result is then divided by 2, regardless of the value of `n`. 

My question is: Why is the division fixed at 2? Shouldn't the result be divided by `n` (i.e., the number of elements summed) if we are truly averaging? Any clarification on this be greatly appreciated! Thanks in advance!

**Comment by jhoydis** [comment]
Hi @LiPtP0000,

It is a long time that I have written this code, but I think that what you are missing is the part that comes before:

```py
        # (Optional) Time-averaging across adjacent DMRS OFDM symbols
        if self._dmrs_length==2:
            # Reshape last dim to [num_dmrs_syms, num_pilots_per_dmrs_sym]
            h_hat = split_dim(h_hat, [self._num_dmrs_syms,
                                      self._num_pilots_per_dmrs_sym], 5)

            # Average adjacent DMRS symbols in time domain
            h_hat = (h_hat[...,0::2,:]+h_hat[...,1::2,:]) \
                     / tf.cast(2, h_hat.dtype)
            h_hat = tf.repeat(h_hat, 2, axis=-2)
            h_hat = tf.reshape(h_hat, h_ls_shape)

            # The error variance gets reduced by a factor of two
            err_var /= tf.cast(2, err_var.dtype)

        # Frequency-averaging between adjacent channel estimates

        # Compute number of elements across which frequency averaging should
        # be done. This includes the zeroed elements.
        n = 2*self._num_cdm_groups_without_data
        k = int(h_hat.shape[-1]/n) # Second dimension

        # Reshape last dimension to [k, n]
        h_hat = split_dim(h_hat, [k, n], 5)
        cond = tf.abs(h_hat)>0 # Mask for irrelevant channel estimates
        h_hat = tf.reduce_sum(h_hat, axis=-1, keepdims=True) \
                / tf.cast(2,h_hat.dtype)
        h_hat = tf.repeat(h_hat, n, axis=-1)
        h_hat = tf.where(cond, h_hat, 0) # Mask irrelevant channel estimates
        h_hat = tf.reshape(h_hat, h_ls_shape)

        # The error variance gets reduced by a factor of two
        err_var /= tf.cast(2, err_var.dtype)
```

You average over two adjacent channel estimates in frequency and (optionally) over two adjacent estimates in the time domain.
For `n=4`, you divide the variance of the error twice by two. Hope that explains it.

**Comment by LiPtP0000** [comment]
Thanks for your reply. However, if `num_cdm_groups_without_data` varies, should the **divisor** of  frequency averaging operation change respectively?

Consider a double-symbol type 2 DMRS scheme with `num_cdm_groups_without_data` set to 3, which means all the CDM groups will transmit DMRS, i.e. the pilots. We focus on single Tx-Rx pair and a single stream, i.e. the **pilot** axis. In the code we first average the estimates of pilots at time domain, then reshape it to original shape. Now the pilot axis may be:
`h1 h1 h2 h2 h3 h3 …`
 But on frequency domain, These 6 pilots are summed up and then divided by 2, then the final estimate is:
`(h1 + h2 + h3) / 2`
Why is it reasonable? I am not quite understand.

I am also confused about the physical meanings of `num_rx` and `num_rx_ant`. Is the latter representing numbers of physical antennas and the former representing antenna ports?

**Reply by jhoydis** [reply]
Hi @LiPtP0000,

You are right that the channel estimates are summed up, but (!) only two of them are nonzero. There is also the comment in the code "This includes the zeroed elements."

 The reason is that the CDM groups do not interfere between each other because they are put on non-overlapping subcarriers (see, e.g., Cell 20 in this [tutorial notebook](https://nvlabs.github.io/sionna/phy/tutorials/5G_NR_PUSCH.html#Understanding-the-DMRS-Configuration)).

In summary, there are only two adjacent DMRS in the frequency domain and (optionally) two adjacent DRMS in the time domain for every port across which you average. You could simply inspect the values of `h_hat` and convince yourself that this is true.

Concerning the second part of your question, `num_rx` refers to the number of receivers in your system. In the downlink, this would be the number of users and in the uplink the number of base stations. The value `num_rx_ant` refers to the number of antennas at each receiver. For example, `num_rx=4` and `num_rx_ant=2` could describe a system in the downlink with for users, where each of them has two antennas. You have the equivalent for `num_tx` and `num_tx_ant`. 

Hope that helps to clarify your issue.

**Reply by LiPtP0000** [reply]
It helps a lot for me! Thanks in advance.

So in practice, although an entire OFDM symbol may be reserved for DMRS, each antenna port—corresponding to a stream after layer mapping—actually occupies **only two REs** in the resource grid on frequency domain. The two ports are further integrated into the two REs through the CDM, forming the CDM group. In the frequency-domain perspective, we can do frequency averaging like the code because other REs have zero estimates when **considering a specific antenna port.**

Nevertheless, the CDM group can be extended in time domain by applying time-domain OCC when double-symbol DMRS is configured, where it contains DMRS symbols *up to* four antenna ports. In this case, an antenna port will occupy **four REs**, so we need to do time averaging then frequency averaging to get a fair estimate.

I hope my understanding is correct this time.

**Reply by jhoydis** [reply, marked as answer]
Yes, this is almost correct. Keep in mind that there are far more pilots per port across a full OFDM symbol than two. 
You either have pilots on every other subcarrier (i.e. six pilots per PRB) or two adjacent subcarriers every six subcarriers (i.e., 4 pilots per PRB). 

Here is the relevant extract from the tutorial that it explains this:

<img width="872" height="419" alt="image" src="https://github.com/user-attachments/assets/62698ae1-6c56-43df-8875-444db7cfb125" />

**Reply by LiPtP0000** [reply]
I think I understand now. Thanks a lot for your clarifications!
